src/stats/stats.jl

```diff
@@ -0,0 +1,165 @@
+export AbstractExecutionStats, GenericExecutionStats,
+       AbstractExtraStats, DictExtraStats,
+       set_stat!, get_stat, statshead, statsline
+
+const STATUSES = Dict(:unknown => "unknown",
+                      :first_order => "first-order stationary",
+                      :max_eval => "maximum number of function evaluations",
+                      :max_time => "maximum elapsed time",
+                      :max_iter => "maximum iteration",
+                      :neg_pred => "negative predicted reduction",
+                      :unbounded => "objective function may be unbounded from below",
+                      :exception => "unhandled exception",
+                      :stalled => "stalled"
+                     )
+
+abstract type AbstractExecutionStats end
+
+abstract type AbstractExtraStats <: AbstractExecutionStats end
+
+type GenericExecutionStats <: AbstractExecutionStats
+  status :: Symbol
+  solution :: Vector # x
+  objective :: Float64 # f(x)
+  dual_feas :: Float64 # ‖∇f(x)‖₂ for unc, ‖P[x - ∇f(x)] - x‖₂ for bnd, etc.
+  iter :: Int
+  counters :: NLPModels.Counters
+  elapsed_time :: Float64
+  solver_specific :: AbstractExtraStats
+end
+
+function GenericExecutionStats(status :: Symbol,
+                           nlp :: AbstractNLPModel;
+                           solution :: Vector=Float64[],
+                           objective :: Float64=Inf,
+                           dual_feas :: Float64=Inf,
+                           iter :: Int=-1,
+                           elapsed_time :: Float64=Inf,
+                           solver_specific :: AbstractExtraStats=DictExtraStats())
+  if !(status in keys(STATUSES))
+    s = join(keys(STATUSES, ", "))
+    error("$status is not a valid status. Use one of the following: $s")
+  end
+  c = Counters()
+  for counter in fieldnames(Counters)
+    setfield!(c, counter, eval(parse("$counter"))(nlp))
+  end
```

Comment:
Can't we use `copy(nlp.counters)` here?

Reply:
Due to models like `SlackModel`, `nlp.counters` may fail. We could add a function `counters` or `copy_counters` to `NLPModels`. We actually had `counters(nlp)`, but using it to get individual counters was ugly. We could add it again, but only use it here.

Reply:
@dpo What do you think here? Add `counters(nlp)` on `NLPModels` again?

Reply:
Would it be bad to leave `counters` out of the stats type? We have methods to access the counters for each type of problem. It seems difficult to include them in stats in a generic way. Perhaps solvers could return something like `(stats, nlp.counters)` or, in the case of a SlackModel, `(stats, nlp.model.counters)`. What do you think? It's not perfect either.

Reply:
If `counters` is out of stats then `solve_problems` has to do the job in a similar way ([here](https://github.com/JuliaSmoothOptimizers/Optimize.jl/blob/master/src/bmark/run_solver.jl#L112)). More specifically, the return, which ends up calling the counters one by one.

> Perhaps solvers could return something like (stats, nlp.counters) or, in the case of a SlackModel, (stats, nlp.model.counters). What do you think? It's not perfect either

That would require checking inside the solver the type of `nlp`, and wouldn't work with user-defined NLPs, unless we create `counters(nlp)`.